dockerclient/client.go

```diff
@@ -578,24 +578,33 @@ func (e *ClientExecutor) LoadImage(from string) (*docker.Image, error) {
 	}
 	for _, config := range auth {
 		// TODO: handle IDs?
-		pullWriter := imageprogress.NewPullWriter(outputProgress)
-		defer pullWriter.Close()
-
-		pullImageOptions := docker.PullImageOptions{
-			Repository:    repository,
-			Tag:           tag,
-			OutputStream:  pullWriter,
-			RawJSONStream: true,
-		}
-		if glog.V(5) {
-			pullImageOptions.OutputStream = os.Stderr
-			pullImageOptions.RawJSONStream = false
-		}
-		authConfig := docker.AuthConfiguration{Username: config.Username, ServerAddress: config.ServerAddress, Password: config.Password}
-		if err = e.Client.PullImage(pullImageOptions, authConfig); err == nil {
+		var pullErr error
+		func() { // A scope for defer
+			pullWriter := imageprogress.NewPullWriter(outputProgress)
+			defer func() {
+				err := pullWriter.Close()
+				if pullErr == nil {
+					pullErr = err
+				}
+			}()
+
+			pullImageOptions := docker.PullImageOptions{
+				Repository:    repository,
+				Tag:           tag,
+				OutputStream:  pullWriter,
+				RawJSONStream: true,
+			}
+			if glog.V(5) {
+				pullImageOptions.OutputStream = os.Stderr
+				pullImageOptions.RawJSONStream = false
+			}
+			authConfig := docker.AuthConfiguration{Username: config.Username, ServerAddress: config.ServerAddress, Password: config.Password}
+			pullErr = e.Client.PullImage(pullImageOptions, authConfig)
+		}()
+		if pullErr == nil {
 			break
 		}
-		lastErr = err
+		lastErr = pullErr
```

Comment:
Looks like you're right and as you noted in your other comment, it's pre-existing.  My first thought was to print each error out to the debug log, but it doesn't look like we're at least currently writing to one from this code.  We could potentially write all the errors out but if the pull fails and then the second auth works would you want to see an error in that instance?   hmmmm....

Reply:
You are right, and this could be confusing. OTOH:

- It is pre-existing; this PR changes nothing about it.
- The way `lastErr` is used just below, the error returned by `LoadImage` is untyped either way (and I’m not …too… motivated to care about any potential callers who behave differently based on an error string)
- `auth` is, technically, a list, but the only caller who sets it to something non-trivial is `openshift/origin/pkg/build/builder/buildDirectImage`, which creates it from a `map[registry]…`, ultimately from `DockerBuilder.setupPullSecret`. The `map[registry]…` format does not make it _easy_ to list multiple credentials for a single registry — although, because `BasicDockerKeyring` does make some normalization of the map key, it probably can happen in rare cases — and it does not make all that much sense for the user to set up a `PullSecret` which is ambiguous about what credentials to use for a single registry.